### ISSUE TITLE
[native_assets_cli] Access to dart_api.h and dart_api_dl.h + passing in the api version

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/native#839]
When compiling native code from source, it would be nice if we'd compile against `dart_api.h` or `dart_api_dl.h` from the Dart/Flutter SDK that is invoking `build.dart`. This would prevent version skew between the native API and the Dart VM.

It would not prevent version skew with precompiled binaries, as they already have been compiled.

To prevent version skew with precompiled binaries we'd need to pass `dart_api_version.h`'s version as an argument as well, so that the users can precompile a version per Dart API. This would be analogous to what python is doing. (@brianquinlan You pointed to examples of this earlier, but I can't seem to find where you did that.) Note that at this point you'd need an older Dart SDK to precompile a dylib to an older API version.

**Comment by brianquinlan** [comment]
> To prevent version skew with precompiled binaries we'd need to pass `dart_api_version.h`'s version as an argument as well, so that the users can precompile a version per Dart API. This would be analogous to what python is doing. (@brianquinlan You pointed to examples of this earlier, but I can't seem to find where you did that.) Note that at this point you'd need an older Dart SDK to precompile a dylib to an older API version.

I think that I showed you the downloadable files associated with a Python package, e.g. for [tensorflow](https://pypi.org/project/tensorflow/#files).

You can see that the names are like "tensorflow-2.15.0-cp311-cp311-macosx_12_0_arm64.whl"

So this is the file that will be downloaded if you install tensorflow 2.15 for Python 3.11.x on a ARM64 mac. 0.0.x version changes never break ABI compatibility but all other version changes are assumed to break ABI compatibility.

There is a limited API that is guaranteed to be ABI compatible for all of Python 3.x.x (3.0 was released 15 years ago and there are still no plans for a 4.0).

I'm not advocating for this, just explaining how it works ;-)

I would advocate for allowing some sort of binary packaging so that users don't need a compiler to use packages containing native code.

**Comment by dcharkes** [comment]
> I would advocate for allowing some sort of binary packaging so that users don't need a compiler to use packages containing native code.

Yes, that is totally possible with the native assets feature. (And passing the dart_api_version would keep the precompiled capabilities on par with the on-demand-compiled capabilities when adding the path to dart_api_dl.h and friends.)

**Comment by dcharkes** [comment]
Preventing version skew between `dart_api.h` and generated FFIgen bindings could theoretically be solved by providing access to version in Dart, and then branching between two sets of generated bindings. (This is a similar issue to https://github.com/dart-lang/native/issues/1589.)

To align the API with access to `flutter.h`:

```dart
config
  .code
    .dart
      .cApi
        .headerPath
        .includeDir
        .version
```

This doesn't entirely cover it properly though. We have two APIs in reality:

1. `dart_api.h` is _only_ available on Dart standalone, and not in Flutter. And it's not covered by `version.h`.
2. `dart_api_dl.h` is available in both Dart and Flutter, and it is covered by `version.h`.

And API 1 doesn't really work on Windows when not also building Dart without a workaround: https://dart-review.googlesource.com/c/sdk/+/400582

So maybe

```dart
config
  .code
    .dart
      .cApiDynamicallyLinked
        .headerPath
        .includeDir
        .version
      .cApi
        .headerPath
        .includeDir
```

**Comment by dcharkes** [comment]
> > Preventing version skew between dart_api.h ...
> 
> This is an embedder api intended for C code embedding the Dart VM. It's not intended to be used from native code called by Dart apps. We also reserve the right to change this API at-will. Why would we want to expose this via native assets?

👍 

Maybe we should consider leaning into that and making two separate include directories, so that we don't give users the wrong idea by giving them a `dart_api.h` in the include dir.

**Comment by dcharkes** [comment]
@jonasfj Mentioned we have Dart SDK packages. I'll explore doing it as an SDK package:

* The `include/` dir.
* a `hook/build.dart` that compiles it into a dylib.
* generated bindings with `package:ffigen`.

This mandates that other packages use the dynamic linker to link against it. (Which we should have support for now: https://github.com/dart-lang/native/issues/190)
